File: lib/c/effects/vignette.c
#include "effects/vignette.h"
#include "config.h"
#include "effects/effect.h"
#include "fastmath.h"

// =================================================================================================
// Default Values
// =================================================================================================

// Default background grey level: 35/255 in sRGB space
#define VIGNETTE_DEFAULT_BACKGROUND 0.1372549f

// Default vignette strength: 40% max darkening at corners
#define VIGNETTE_DEFAULT_STRENGTH 0.4f

// =================================================================================================
// Effect Implementation
// =================================================================================================

void effect_vignette_apply(float *fb, int width, int height, const void *config,
                           const void *cache) {
  // If no config or geometry, nothing to do
  if (!config || !cache)
    return;

  const VignetteConfig *cfg = (const VignetteConfig *)config;

  // Skip if vignette disabled
  if (!cfg->enabled)
    return;

  const VignetteGeometry *geom = (const VignetteGeometry *)cache;

  // Get configuration values with defaults
  // strength: negative means use default, 0 means no darkening, positive means use that value
  // background: negative or zero means use default, positive means use that value
  float strength = cfg->strength >= 0.0f ? cfg->strength : VIGNETTE_DEFAULT_STRENGTH;
  float bg_base = cfg->background > 0.0f ? cfg->background : VIGNETTE_DEFAULT_BACKGROUND;

  // Circle bounds
  float cx = geom->cx;
  float cy = geom->cy;
  float radius = geom->radius;
  float r2 = radius * radius;

  // Maximum distance from center (diagonal to corner)
  float max_dist = sqrtf_impl((float)(width * width + height * height)) * 0.5f;

  for (int y = 0; y < height; y++) {
    float dy = (float)y - cy;
    float dy2 = dy * dy;
    int row_offset = y * width * 4;

    for (int x = 0; x < width; x++) {
      float dx = (float)x - cx;
      float dist2 = dx * dx + dy2;

      // Only process pixels OUTSIDE the watch circle
      if (dist2 > r2) {
        int idx = row_offset + x * 4;

        // Calculate vignette factor based on distance from center
        float dist_from_center = sqrtf_impl(dist2);

        // Normalized distance: 0.0 at circle edge, 1.0 at max distance
        float vignette_t = clampf((dist_from_center - radius) / (max_dist - radius), 0.0f, 1.0f);

        // Smoothstep for perceptually smoother gradient
        // smoothstep(t) = t^2 * (3 - 2t)
        float smooth_t = vignette_t * vignette_t * (3.0f - 2.0f * vignette_t);

        // Vignette darkening factor: 1.0 at edge, (1.0 - strength) at max distance
        float vignette = 1.0f - smooth_t * strength;

        // Apply dithering noise to break up banding in dark gradients
        uint32_t hash = vignette_hash_pixel(x, y);
        // Noise in range [-1, +1], scaled to ~1 unit in 0-255 space (±0.5/255)
        float dither = ((float)(hash & 0xFF) / 255.0f - 0.5f) * (2.0f / 255.0f);

        // Final grey value with vignette and dither
        float grey = clampf(bg_base * vignette + dither, 0.0f, 1.0f);

        // Set pixel to grey background with full opacity
        fb[idx] = grey;
        fb[idx + 1] = grey;
        fb[idx + 2] = grey;
        fb[idx + 3] = 1.0f;
      }
    }
  }
}

// Effect descriptor
const Effect EFFECT_VIGNETTE = {.name = "vignette", .apply = effect_vignette_apply};

File: lib/zig/effects/vignette.zig
const std = @import("std");

const color = @import("../color/color.zig");
const grain = @import("grain.zig");

const default_background: f32 = 0.1372549; // 35/255 in sRGB
const default_strength: f32 = 0.4; // 40% max darkening at corners

/// Vignette effect configuration.
pub const Config = struct {
    enabled: bool = true,
    strength: f32 = default_strength,
    background: f32 = default_background,
};

/// Geometry context for vignette (watch circle bounds).
pub const Geometry = struct {
    center_x: f32,
    center_y: f32,
    radius: f32,
};

/// Apply vignette effect to buffer background (outside watch circle).
/// Expects buffer to be in sRGB space.
pub fn apply(
    buffer: []color.Color,
    width: usize,
    height: usize,
    config: Config,
    geometry: Geometry,
) void {
    if (!config.enabled) return;

    const strength = if (config.strength >= 0.0) config.strength else default_strength;
    const bg_base = if (config.background > 0.0) config.background else default_background;

    const cx = geometry.center_x;
    const cy = geometry.center_y;
    const radius = geometry.radius;
    const r2 = radius * radius;

    // Maximum distance from center (diagonal to corner)
    const width_f: f32 = @floatFromInt(width);
    const height_f: f32 = @floatFromInt(height);
    const max_dist = @sqrt(width_f * width_f + height_f * height_f) * 0.5;
    const inv_dist_range = 1.0 / (max_dist - radius);

    for (0..height) |y| {
        const dy = @as(f32, @floatFromInt(y)) - cy;
        const dy2 = dy * dy;

        for (0..width) |x| {
            const dx = @as(f32, @floatFromInt(x)) - cx;
            const dist2 = dx * dx + dy2;

            // Only process pixels OUTSIDE the watch circle
            if (dist2 > r2) {
                const idx = y * width + x;

                // Calculate vignette factor based on distance from center
                const dist_from_center = @sqrt(dist2);

                // Normalized distance: 0.0 at circle edge, 1.0 at max distance
                const vignette_t = std.math.clamp((dist_from_center - radius) * inv_dist_range, 0.0, 1.0);

                // Smoothstep for perceptually smoother gradient
                const smooth_t = vignette_t * vignette_t * (3.0 - 2.0 * vignette_t);

                // Vignette darkening factor: 1.0 at edge, (1.0 - strength) at max distance
                const vignette_factor = 1.0 - smooth_t * strength;

                // Apply dithering noise to break up banding in dark gradients
                const xi: i32 = @intCast(x);
                const yi: i32 = @intCast(y);
                const hash = grain.hashPixel(xi, yi);
                // Noise in range [-1, +1], scaled to ~1 unit in 0-255 space (±0.5/255)
                const dither = (@as(f32, @floatFromInt(hash & 0xFF)) / 255.0 - 0.5) * (2.0 / 255.0);

                // Final grey value with vignette and dither
                const grey = std.math.clamp(bg_base * vignette_factor + dither, 0.0, 1.0);

                // Set pixel to grey background with full opacity
                buffer[idx] = color.rgba(grey, grey, grey, 1.0);
            }
        }
    }
}
